### PR TITLE
osd: replace hb_out and hb_in with a single hb_peers

### PR DESCRIPTION
**Pull request by liupan1111**
Signed-off-by: Pan Liu <pan.liu@istuary.com>

http://tracker.ceph.com/issues/18057

### CODE REVIEW

**Comment by liupan1111** [comment]
@liewegas , Please help take a look. Thanks.

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -759,8 +759,8 @@ void OSDService::update_osd_stat(vector<int>& hb_peers)
 {
   Mutex::Locker lock(stat_lock);
 
-  osd_stat.hb_in.swap(hb_peers);
-  osd_stat.hb_out.clear();
+  osd_stat.hb_out.swap(hb_peers);
```
Comment (liupan1111) [inline src/osd/OSD.cc]: @tchaikov. First, it is correct, update_osd_stat() is only called by OSD::heartbeat, and hb_peers is an input parameter for lupdate_osd_stat. 

Let's take a look how is "hb_peers" used in OSD::heartbeat: "hb_peers" is got from heartbeat_peers, and heart beat will be sent to these peers one by one from current osd by using send_message.

In this way, these hb_peers are hb_out, not hb_in. That is what I want to fix in this PR.

Then for the issue 2896. As I described above, "hb_out" is what we want to display, and "hb_in" is vestigal.

Comment (tchaikov) [inline src/osd/OSD.cc]: oic, the `hb_out` stands for "peers for outgoing heartbeat messages".

Comment (tchaikov) [inline src/osd/OSD.cc]: can we take this chance to remove `hb_in` in a separated commit?

@liewegas what do you think?

**Comment by liewegas** [comment]
Yeah, I'd rather remove it (and possiblye rename the remaining field hb_peers or something).

**Comment by liupan1111** [comment]
sure, i will remove hb_in and rename hb_out in seperate commits in this PR.

**Comment by liupan1111** [comment]
@liewegas @tchaikov  The modification is done.

File: src/osd/osd_types.cc [hunk 2]
```diff
@@ -335,8 +331,7 @@ void osd_stat_t::encode(bufferlist &bl) const
   ::encode(kb_avail, bl);
   ::encode(snap_trim_queue_len, bl);
   ::encode(num_snap_trimming, bl);
-  ::encode(hb_in, bl);
-  ::encode(hb_out, bl);
+  ::encode(hb_peers, bl);
```
Comment (liewegas) [inline src/osd/osd_types.cc]: we can't drop a field or else we break compatibility.  hb_peers replaces hb_in.  in place of hb_out, ::encode((uint32_t)0, bl);

File: src/osd/osd_types.cc [hunk 3]
```diff
@@ -350,8 +345,7 @@ void osd_stat_t::decode(bufferlist::iterator &bl)
   ::decode(kb_avail, bl);
   ::decode(snap_trim_queue_len, bl);
   ::decode(num_snap_trimming, bl);
-  ::decode(hb_in, bl);
-  ::decode(hb_out, bl);
+  ::decode(hb_peers, bl);
```
Comment (liewegas) [inline src/osd/osd_types.cc]: uint32_t num_hb_out;
::deocde(num_hb_out, bl);

File: src/osd/osd_types.cc (hunk 2 above)
Comment (liewegas) [inline src/osd/osd_types.cc]: also, update the encoding version (ENCODE_START(5, 2, bl)) just in case we need to distinguish between the encodings.

Comment (liewegas) [inline src/osd/osd_types.cc]: change the 4 to 5 in the DECODE_START_.. macro in decode() too

File: src/osd/osd_types.cc [hunk 4]
```diff
@@ -329,29 +325,31 @@ void osd_stat_t::dump(Formatter *f) const
...
+  ::encode((uint32_t)0, bl);
   ::encode(op_queue_age_hist, bl);
   ::encode(fs_perf_stat, bl);
   ENCODE_FINISH(bl);
 }
 
 void osd_stat_t::decode(bufferlist::iterator &bl)
 {
-  DECODE_START_LEGACY_COMPAT_LEN(4, 2, 2, bl);
+  uint32_t num_hb_out;
```
Comment (tchaikov) [inline src/osd/osd_types.cc]: nit, could just move this line to where it is firstly used (line:352)

**Comment by yuriw** [comment]
This PR was part of this tests http://pulpito.ceph.com/yuriw-2016-11-28_20:50:22-rados-wip-yuri-testing_2016_11_28-distro-basic-smithi/

@liewegas - pls merge

**Pull request merged**